src/marin/processing/classification/autoscaler.py

```diff
@@ -0,0 +1,456 @@
+        for _ in range(NUM_RETRIES):
+            try:
+                actor = self.actor_class.options(**self.actor_options).remote(
+                    self.model_name_or_path, self.attribute_name, self.model_type, **self.actor_kwargs
+                )
+                break
+            except Exception as e:
+                logger.warning(f"Actor creation failed: {e}")
+                time.sleep(1)
+        try:
+            # Ensure the actor is scheduled and ready before exposing to dispatch
+            ray.get(actor.ping.remote())
+        except Exception as e:
+            logger.warning(f"Actor readiness check failed: {e}")
+        self.actors.append(actor)
+        self.actor_futures[actor] = []
+
+    def _start_autoscaling_monitor(self):
+        """Start a local background thread for monitoring and autoscaling.
+
+        Avoids passing the non-serializable pool object to a Ray actor.
+        """
+        self._monitor_stop_event: Event = Event()
+
+        def _loop():
+            while not self._monitor_stop_event.is_set():
+                try:
+                    time.sleep(self.scale_check_interval)
+                    self._check_and_scale()
+                except Exception:
+                    # Best-effort; do not crash the monitor
+                    pass
+
+        self._monitor_thread: Thread = Thread(target=_loop, daemon=True)
+        self._monitor_thread.start()
+
+    def _start_dispatcher(self):
+        """Start a background dispatcher that pulls tasks and dispatches work."""
+        self._dispatch_stop_event: Event = Event()
+
+        def _loop():
+            while not self._dispatch_stop_event.is_set():
+                if len(self.actors) == 0:
+                    logger.debug("Dispatcher waiting for actors to come online")
+                    time.sleep(0.5)
+                    continue
+
+                dispatched_any = False
+                for _ in range(4):  # cap per-iteration dispatch volume
+                    try:
+                        task = self.task_queue.get(timeout=1)
+                    except Exception:
+                        break
+
+                    self.total_submitted += 1
+                    try:
+                        self._dispatch_task(task)
+                        dispatched_any = True
+                    except RuntimeError:
+                        # No actor available; put task back and wait before retrying
+                        try:
+                            self.task_queue.put(task)
+                        except Exception:
+                            logger.error("Dispatcher failed to requeue task after no actor available")
+                        time.sleep(0.1)
+                        break
+
+                if not dispatched_any:
+                    logger.info("No tasks dispatched - going to sleep for 10 seconds.")
+                    time.sleep(10)
+
+        self._dispatch_thread: Thread = Thread(target=_loop, daemon=True)
```

Comment:
it's a little weird to define the thread & stop_event here, you can just create hem in init?